lib/accept-constraints.js

```diff
@@ -0,0 +1,49 @@
+'use strict'
+
+const ConstraintsStore = require('./constraints-store')
+
+const acceptVersionStrategy = require('./strategies/accept-version')
+const acceptHostStrategy = require('./strategies/accept-host')
+
+const DEFAULT_STRATEGIES_NAMES = ['version', 'host']
+
+module.exports = (customStrategies) => {
+  const strategies = [
+    new acceptVersionStrategy(),
+    new acceptHostStrategy()
+  ]
+
+  if (customStrategies) {
+    for (let i = 0; i < customStrategies.length; i++) {
+      const strategy = new customStrategies[i]()
+      if (DEFAULT_STRATEGIES_NAMES.indexOf(strategy.name) !== -1) {
+        strategies[i] = strategy
+      } else {
+        strategies.push(strategy)
+      }
+    }
+  }
+
+  return {
+    storage: function () {
+      const stores = {}
+      for (var i = 0; i < strategies.length; i++) {
+        stores[strategies[i].name] = strategies[i].storage()
+      }
+      return ConstraintsStore(stores)
+    },
+    deriveConstraints: function (req, ctx) {
+      const derivedConstraints = {}
```

Comment:
This is no longer a super hot path and only gets invoked once per route match, right? Maybe not worth it, but if you wanted to be really fancy, you could set up this object's prototype outside this object literal with the early-prototype-definition trick Fastify uses elsewhere and that you mentioned in a comment.

```javascript
 module.exports = (customStrategies) => {
  const strategies = [
    new acceptVersionStrategy(),
    new acceptHostStrategy()
  ]

  if (customStrategies) {
    for (let i = 0; i < customStrategies.length; i++) {
      const strategy = new customStrategies[i]()
      if (DEFAULT_STRATEGIES_NAMES.indexOf(strategy.name) !== -1) {
        strategies[i] = strategy
      } else {
        strategies.push(strategy)
      }
    }
  }

  // Bang out a full prototype ahead of time
  const derivedConstraintsPrototype = class {};
  for (const strategy of strategies) {
    derivedConstraintsPrototype.prototype[strategy.name] = null
  }

  return {
    storage: function () {
      const stores = {}
      for (var i = 0; i < strategies.length; i++) {
        stores[strategies[i].name] = strategies[i].storage()
      }
      return ConstraintsStore(stores)
    },
    deriveConstraints: function (req, ctx) {
     const derivedConstraints = new derivedConstraintsPrototype()
     let value, hasConstraint = false
      for (var i = 0; i < strategies.length; i++) {
        value = strategies[i].deriveConstraint(req, ctx)
        if (value) {
          hasConstraint = true
          derivedConstraints[strategies[i].name] = value
        }
      }

      return hasConstraint ? derivedConstraints : null
    }
  }
}
```